pkg/pillar/cmd/nodeagent/handletimers.go

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) 2017-2018 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package nodeagent
+
+import (
+	"fmt"
+	"syscall"
+	"time"
+
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	"github.com/lf-edge/eve/pkg/pillar/zboot"
+	log "github.com/sirupsen/logrus"
+)
+
+// node health timer funcions
+
+// called periodically from agentwatch context
+func handleDeviceTimers(ctx *nodeagentContext) {
+	handleFallbackOnCloudDisconnect(ctx)
+	handleResetOnCloudDisconnect(ctx)
+}
+
+// when baseos upgrade is in progress
+// on cloud disconnect for a specified amount of time, reset the node
+func handleFallbackOnCloudDisconnect(ctx *nodeagentContext) {
+	if !ctx.updateInprogress {
+		return
+	}
+	// apply the fallback time function
+	fallbackLimit := time.Second * time.Duration(globalConfig.FallbackIfCloudGoneTime)
+	timePassed := time.Since(ctx.lastConfigReceivedTime)
+	if timePassed > fallbackLimit {
+		errStr := fmt.Sprintf("Exceeded fallback outage for cloud connectivity %d by %d seconds; rebooting\n",
+			fallbackLimit/time.Second,
+			(timePassed-fallbackLimit)/time.Second)
+		log.Errorf(errStr)
+		execReboot(ctx, errStr)
+	}
+}
+
+// on cloud disconnect for a specified amount time, reset the node
+func handleResetOnCloudDisconnect(ctx *nodeagentContext) {
+	// apply the reset time function
+	resetLimit := time.Second * time.Duration(globalConfig.ResetIfCloudGoneTime)
+	timePassed := time.Since(ctx.lastConfigReceivedTime)
+	if timePassed > resetLimit {
+		errStr := fmt.Sprintf("Exceeded outage for cloud connectivity %d by %d seconds; rebooting\n",
+			resetLimit/time.Second,
+			(timePassed-resetLimit)/time.Second)
+		log.Errorf(errStr)
+		execReboot(ctx, errStr)
+	}
+}
+
+// baseos upgrade test validation utilities
+func setTestStartTime(ctx *nodeagentContext) {
+	// test has already started or, the current
+	// partition is not in inprogress state,
+	// nothing to do
+	if !ctx.updateInprogress || ctx.testInprogress {
+		return
+	}
+	// only when current partition state is in progress
+	// start the test
+	ctx.upgradeTestStartTime = time.Now()
```

Comment:
This can be problematic if the hardware doesn't have a time of day chip + battery since we don't run ntp until after we've started nodeagent. Using a normal timer as in zedagent is robust against the ToD changing.

Reply:
Have simplified the time function. It is not going to use ToD value.